### PR TITLE
[WIP] Curb your round end

### PR DESCRIPTION
**Pull request by ghost**
https://youtu.be/eIoAVm6-ZkY
:cl:
* rscadd: round end now has a credits sequence inspired by Curb Your Enthusiasm

### CODE REVIEW

**Comment by ghost** [comment]
uh, "ported" (the idea and like ~50%? of the code) from Bay, https://github.com/Baystation12/Baystation12/pull/18348

File: code/game/rollcredits.dm [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		if(H.timeofdeath < 5 MINUTES) //no prespawned corpses
+			continue
+		else if(H.real_name)
+			corpses += H.real_name
+	if(corpses.len)
+		titles += "<center>BASED ON REAL EVENTS<br>In memory of [english_list(corpses)].</center>"
+
+	var/list/staff = list("<h1>PRODUCTION STAFF</h1>")
+	var/list/staffjobs = list("Coffee Fetcher", "Cameraman", "Angry Yeller", "Chair Operator", "Choreographer", "Historical Consultant", "Costume Designer", "Chief Editor", "Executive Assistant")
+	for(var/client/C)
```
Comment (MadmanMartian) [inline code/game/rollcredits.dm]: Is there no list of admins that can be used instead?

File: code/game/rollcredits.dm [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+			continue
+		if(C.holder.rights & (R_DEBUG|R_ADMIN))
+			var/observername = ""
+			if(C.mob && istype(C.mob,/mob/dead/observer))
+				var/mob/dead/observer/O = C.mob
+				if(O.started_as_observer)
+					observername = "[O.real_name] a.k.a. "
+			staff += "<h2>[uppertext(pick(staffjobs))] - [observername]'[C.key]'</h2>"
+
+	var/disclaimer = "Sponsored by The United States Navy.<br>All rights reserved.<br>"
```
Comment (MadmanMartian) [inline code/game/rollcredits.dm]: No we're not.

File: code/game/rollcredits.dm [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	disclaimer += pick("All stunts were performed by underpaid and expendable interns. Do NOT try at home.", "Nanotrasen does not endorse behaviour depicted. Attempt at your own risk.<br>")
+	titles += "<center>[disclaimer]</center>"
+	titles += "<center>This motion picture is protected under the copyright laws of the United States and other countries throughout the world. Country of first publication: United States of America. Any unauthorized exhibition, distribution, or copying of this film or any part thereof (including soundtrack) may result in civil liability</center>"
```
Comment (MadmanMartian) [inline code/game/rollcredits.dm]: No it's not

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 4]
```diff
@@ -94,4 +94,5 @@
 	var/show_client_status_on_examine = TRUE //If false, don't display catatonic/braindead messages to non-admins
 
 	var/become_zombie_after_death = FALSE
-	var/times_cloned = 0 //How many times this person has been cloned
\ No newline at end of file
+	var/times_cloned = 0 //How many times this person has been cloned
+	var/talkcount = 0
```
Comment (MadmanMartian) [inline code/modules/mob/living/carbon/human/human_defines.dm]: Put a comment next to this describing what the hell it is and what it's used for.

File: code/game/rollcredits.dm (hunk 2 above)
Comment (ghost) [reply code/game/rollcredits.dm]: Yeah we are

File: code/game/rollcredits.dm (hunk 1 above)
Comment (ghost) [reply code/game/rollcredits.dm]: Is there?

File: code/modules/mob/living/carbon/human/human_defines.dm (hunk 4 above)
Comment (ghost) [reply code/modules/mob/living/carbon/human/human_defines.dm]: Aight

File: code/game/rollcredits.dm (hunk 3 above)
Comment (ghost) [reply code/game/rollcredits.dm]: https://en.wikipedia.org/wiki/All_persons_fictitious_disclaimer

**Review by MadmanMartian — COMMENTED** [review]
Suggestion: Have a check to see if the pets are alive, and then in the 'legals' bit have an inline if statement for if animals were or weren't harmed during the making of this film.

File: code/game/rollcredits.dm (hunk 3 above)
Comment (MadmanMartian) [reply code/game/rollcredits.dm]: It's not protected under the copyright laws of the united states. This is stuff that can get weird if somebody decides to chase up on it

File: code/game/rollcredits.dm (hunk 2 above)
Comment (MadmanMartian) [reply code/game/rollcredits.dm]: We're not sponsored by the United States Navy. This is stuff that can get weird if somebody decides to chase up on it

**Comment by ghost** [comment]
@MadmanMartian pretty good idea

**Comment by ghost** [comment]
The reason this PR is up and WIP is I'd like suggestions of what else to put in the credits.
For example, thanks to Arthur if there are no admins online "Alan Smithee" ( https://en.wikipedia.org/wiki/Alan_Smithee ) will be credited as director.

**Comment by utilitas-** [comment]
All persons (theoretically) fictitious.

Exactly X Y were Zed in the making of this film. Could be anything from clowns harmed through plasma mined to bananas peeled. Canisters vented, moles of gas burned, units of chemicals used, amount of blood spilled...

Somewhat random newscaster entries, PDA posts and comms chatter inserted at the end with an accompanying laugh track.

Anyone trying to join but banned being "unfortunately unable to participate due to personal health problems." Similar for those banned within the round.

If you feel like playing around with it, you could take snapshots of events in the round, (for example explosions, framed fights initiated with a gun, someone throwing something heavy, and token custom snapshots taken by admins) crop them and add them as inline 'out-takes' in the credits.

**Comment by ghost** [comment]
When exactly does this play?
At SMELELE? A few seconds after SMELELE? Half a second before disconnecting the players for reboot?

As for shit to cram in it
"Outer Space footage provided by NASA"
"Filmed with LUMMOX © cameras and lenses"
"Based on a True Story"

**Comment by ghost** [comment]
image

**Comment by ghost** [comment]
this post makes me irrationally upset

**Comment by Inorien** [comment]
Red Dwarf credits music and you got yourself a deal

File: code/game/rollcredits.dm [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	if(SNOW_THEME)
+		possible_titles += "A VERY [pick("NANOTRASEN", "EXPEDITIONARY", "DIONA", "PLASMA", "MARTIAN")] CHRISTMAS"
+	possible_titles += "[pick("GUNS, GUNS EVERYWHERE", "MUCH ADO ABOUT NOTHING", "WHAT HAPPENS WHEN YOU MIX MOMMIS AND COMMERCIAL-GRADE PACKING FOAM", "ATTACK! ATTACK! ATTACK!", "SEX BOMB", "THE BALLAD OF [uppertext(station_name())]")]"
+	possible_titles += "[pick("SPACE", "SEXY", "DRAGON", "WARLOCK", "LAUNDRY", "GUN", "ADVERTISING", "DOG", "CARBON MONOXIDE", "NINJA", "WIZARD", "SOCRATIC", "JUVENILE DELIQUENCY", "POLITICALLY MOTIVATED", "RADTACULAR SICKNASTY")] [pick("QUEST", "FORCE", "ADVENTURE")]"
+	possible_titles += "[pick("THE DAY [uppertext(station_name())] STOOD STILL", "HUNT FOR THE GREEN WEENIE", "ALIEN VS VENDOMAT", "SPACE TRACK")]"
+	titles += "<center><h1>SEASON [rand(1,22)] EPISODE [rand(1,17)]<br>[pick(possible_titles)]<h1></h1></h1></center>"
```
Comment (RemieRichards) [inline code/game/rollcredits.dm]: Could make every year after 2003 its own season.

**Comment by ghost** [comment]
/tv/station dream comes true

**Comment by ShiftyRail** [comment]
I saw people complained that it'd get old fast and that it would remove the old stress-relief shuttle grief we're accustomed to.

Suggestion : make this happen in 1 out of every 10 (or maybe less ?) rounds, and have a few checks to see if it's worth rolling the credits (like has the round lasted more than x minutes...)

**Comment by Inorien** [comment]
Make it a traitor journalist thing

**Comment by SangriaMonkey** [comment]
This is kind of neat but the end-of-round killing spree is a time honored tradition.

**Comment by utilitas-** [comment]
What if it triggers upon the emergency shuttle leaving, not when it arrives to CentComm?

Nothing happens on the flight most of the time, anyway, and in case it does, you can make it thematically appropriate. When a canister gets opened, something explodes or someone starts a'shootin' - disc scratch and interrupt the credits until the shuttle arrives at CentComm or everyone is dead, then continue after a snare sound.

**Comment by MadmanMartian** [comment]
Make it an opt-in process.

**Comment by ShiftyRail** [comment]
>Nothing happens on the flight most of the time, anyway

Factually untrue. A good deal of deaths happen midflight because a miner released his pokemons or because someone popped a viscerator. There's also the exchanges between astounded crewmembers, and always at least 2 bodies being defibbed.

**Comment by SkowronX** [comment]
>Sponsored by the United States Navy 

🤔

**Comment by ghost** [comment]
Any chance you can get this to run while the server is rebooting? I.e. replace the now-terrible round-end sounds with this.

**Comment by Blithering** [comment]
It could play on rounds where the shuttle doesn't get called, like malf, rev, blob etc.

**Comment by ShiftyRail** [comment]
@Unknown-as-Captain I don't think so. Round-end sounds work because it's a single resource sent and then played by your client.
This depends on fetching the list of players, admins, etc, and display it. I think it requires a dialogue between client and server, ie it can't be done while the serb is off.

@Blithering good idea, although round-end griff or jokes happens here sometimes.

**Comment by Inorien** [comment]
How about it prepares the file during the flight, and sends the blob resource at the end? So it wouldn't take into account anything that happens during the shuttle as a limitation

**Comment by ghost** [comment]
@ShiftyRail I'm not sure what you mean. The round-end sound is sent to client BEFORE the server shuts down. How would the client be playing them otherwise?

File: code/game/rollcredits.dm [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	var/list/titles = list()
+	var/list/cast = list()
+	var/chunk = "<center>"
+	var/list/possible_titles = list()
+	var/chunksize = 0
+
+	/* Establish a big-ass list of potential titles for the "episode". */
+	possible_titles += "THE [pick("DOWNFALL OF", "RISE OF", "TROUBLE WITH", "FINAL STAND OF", "DARK SIDE OF")] [pick("SPACEMEN", "HUMANITY", "DIGNITY", "SANITY", "THE CHIMPANZEES", "THE VENDOMAT PRICES","[uppertext(station_name())]")]"
+	possible_titles += "THE CREW GETS [pick("RACIST", "PICKLED", "AN INCURABLE DISEASE", "PIZZA", "A VALUABLE HISTORY LESSON", "A BREAK", "HIGH", "TO LIVE", "TO RELIVE THEIR CHILDHOOD", "EMBROILED IN CIVIL WAR", "SERIOUS ABOUT [pick("DRUG ABUSE", "CRIME", "PRODUCTIVITY", "ANCIENT AMERICAN CARTOONS", "SPACEBALL")]")]"
+	possible_titles += "THE CREW LEARNS ABOUT [pick("LOVE", "DRUGS", "THE DANGERS OF MONEY LAUNDERING", "SPACE 'NAM", "INVESTMENT FRAUD", "KELOTANE ABUSE", "RADIATION PROTECTION", "SACRED GEOMETRY", "STRING THEORY", "ABSTRACT MATHEMATICS", "[pick("CATBEAST", "DIONAN", "PLASMAMAN", "VOX", "GREY")] MATING RITUALS", "ANCIENT CHINESE MEDICINE")]"
```
Comment (ghost) [inline code/game/rollcredits.dm]: Lawsets

**Comment by Durgen** [comment]
maybe antag objectives could be incorporated?

**Comment by aceedex** [comment]
wonderful

File: code/game/rollcredits.dm (hunk 6 above)
Comment (ghost) [reply code/game/rollcredits.dm]: good idea

**Comment by ghost** [comment]
@Durgen How do you think they should be?

**Comment by ghost** [comment]
@Unknown-as-Captain Unfortunately because byond is not a great engine, transforms (the credits scrolling) and music stops playing immediately when the server shuts down even though these are client sided effects
The reason (this is a poor reason but thanks byond) that the end round jingle plays even the server dies is because it's sent to the world rather than the client.

@SangriaMonkey The general idea is to have this play just before the server would shut down anyway; there would still be time for round-end grief.

**Comment by ghost** [comment]
Sad but not unexpected

**Comment by Unknown-as-Captain** [comment]
Sad but not unexpected

**Comment by ghost** [comment]
Refactors the code to be much much much nicer, my code was a little hacky and bay's code was a little messy; now it's much cleaner and a little more modular.

adds a skip credits button in the OOC tab
also adds some other suggestions.

credits now play when the server is about to restart - there is still loads of time for shuttle grief

**Comment by ghost** [comment]
oh, and credits can totally be edited by admins during the round if they wanna customise it

**Comment by Kammerjunk** [comment]
>adds a skip credits button in the OOC tab

That's all I'd ask for thanks 👌

**Comment by Durgen** [comment]
@DrCelt 

Costarring

Honky Wonky as Mr. Sigma

- Killed X, Successful

- Stole Y, Successful

- Escaped, Successful

Marty McFly as Mr. Iota

- Killed X, Successful

- Stole Y, Successful

- Escaped, Successful

**Comment by MadmanMartian** [comment]
@DrCelt 
"The winners"
 - Every succesful antag

"The losers"
 - Every unsuccesful antag
 - The clown

**Comment by ghost** [comment]
I don't think this is meant to be a replacement for the round-end info window, nor should it be

**Comment by NigglyWiggly** [comment]
As much as I like this, I feel like trying to have it at the end of every round is a bad idea.
Maybe make it a button admins can trigger at round end? Possibly just a chance of it happening?
Also, I’m guessing this can’t update in real-time, so we can’t see someone go from being listed as alive to being dead if they die during the credits?

**Comment by Sooose** [comment]
Probably controversial: Bans from the round for extreme memeage.

**Comment by Inorien** [comment]
I fear that might attract shitters to do crazy shit just for the purpose of being banned / having their name show up

**Comment by ihadtoregisterforthis** [comment]
How about holiday specific text?

File: code/game/rollcredits.dm [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+/datum/credits/New()
+	var/list/possible_episode_names = list()
+
+	/* Establish a big-ass list of potential titles for the "episode". */
+	possible_episode_names += "THE [pick("DOWNFALL OF", "RISE OF", "TROUBLE WITH", "FINAL STAND OF", "DARK SIDE OF")] [pick("SPACEMEN", "HUMANITY", "DIGNITY", "SANITY", "THE CHIMPANZEES", "THE VENDOMAT PRICES","[uppertext(station_name())]")]"
+	possible_episode_names += "THE CREW GETS [pick("RACIST", "PICKLED", "AN INCURABLE DISEASE", "PIZZA", "A VALUABLE HISTORY LESSON", "A BREAK", "HIGH", "TO LIVE", "TO RELIVE THEIR CHILDHOOD", "EMBROILED IN CIVIL WAR", "SERIOUS ABOUT [pick("DRUG ABUSE", "CRIME", "PRODUCTIVITY", "ANCIENT AMERICAN CARTOONS", "SPACEBALL")]")]"
+	possible_episode_names += "THE CREW LEARNS ABOUT [pick("LOVE", "DRUGS", "THE DANGERS OF MONEY LAUNDERING", "SPACE 'NAM", "INVESTMENT FRAUD", "KELOTANE ABUSE", "RADIATION PROTECTION", "SACRED GEOMETRY", "STRING THEORY", "ABSTRACT MATHEMATICS", "[pick("CATBEAST", "DIONAN", "PLASMAMAN", "VOX", "GREY")] MATING RITUALS", "ANCIENT CHINESE MEDICINE","LAWSETS")]"
+	if(SNOW_THEME)
+		possible_episode_names += "A VERY [pick("NANOTRASEN", "EXPEDITIONARY", "DIONA", "PLASMA", "MARTIAN")] CHRISTMAS"
```
Comment (ghost) [inline code/game/rollcredits.dm:22]: @ihadtoregisterforthis

Comment (ghost) [reply code/game/rollcredits.dm:22]: like this?

**Comment by ihadtoregisterforthis** [comment]
I was thinking about something using [dates.dm](https://github.com/vgstation-coders/vgstation13/blob/491b28e99152f490029981dd8936c2398e4834d7/__DEFINES/dates.dm)

**Comment by ghost** [comment]
@ihadtoregisterforthis how would it look?

**Comment by praisenarsie** [comment]
you could make it so it plays if all antags fail their objectives

**Comment by SysadminJeroen** [comment]
@DrCelt Are you still working on this? Because I still want this.

**Comment by DamianX** [comment]
upon what time

**Comment by DeityLink** [comment]
I fucking love the idea, the implementation not so much. Also does it mean no more post-round griefing/fucking around in centcomm? In which case meh.

**Comment by DamianX** [comment]
he ded

**Pull request closed without merging**